include/lib/map.h

```diff
@@ -4,24 +4,39 @@
 
 #include "stdlib/stdtypes.h"
 
-/** Data structure for a generic map */
+/** Structures for a generic map. */
+struct entry {
+    /** The key associated with the entry. */
+    void *key;
+    /** The value associated with the entry. */
+    void *value;
+};
+
 struct map {
-    //TODO
+    /** Table which holds all entries. */
+    struct entry *table;
+    /** Amount of space used. */
+    size_t used;
+    /** Maximum space to allocate for entries */
+    size_t size;
 };
 
-/** Initializes a new hashtable */
-void map_init(struct map *m);
+/** Initializes a new hashmap. Returns `-1` if not allocated. */
+int map_init(struct map *m, size_t size);
 
-/** Frees the hashtable from memory */
+/** Frees the hashtable from memory. */
 void map_free(struct map *m);
 
-/** Adds the key/value pair to the hashtable */
-void map_put(struct map *m, void *key, void *value);
+/** Adds the key/value pair to the hashtable. */
+int map_put(struct map *m, void *key, void *value, size_t size);
+
+/** Gets the value associated with the specified key. */
+void *map_get(struct map *m, void *key, size_t size);
 
-/** Gets the value associated with the specified key */
-void *map_get(struct map *m, void *key);
+/** Checks if key exists in hashmap. */
+bool map_contains(struct map *m, void *key, size_t size);
 
-/** Returns true if the hashtable contains the key, false otherwise */
-bool map_contains(struct map *m, void *key);
+/** Hash function for strings. */
+size_t fnv_1a(char *str);
```

Comment:
This function should not be declared in the header file. It should be a static function only in the c file. If you think it should be globally accessible, we can create a new compilation unit in the standard library for hash functions. For now, keep it private.

Reply:
I'm making a new stdlib file called stdhash.